texastribune/salesforce-stripe
# Add reason to Salesforce when charge fails

#### What's this PR do?

Updates the opportunity with the reason when a card charge fails. And updates the status to "Closed Lost."

#### Why are we doing this? How does it help us?

Right now we just keep retrying for three days and then give up. The opportunity is left in the `Pledged` state. Then someone has to either manually go through and mark them as `Closed Lost` or run a job to do so. And they don't know why the transaction failed. Sometimes that reason code is useful. These retries have not proved useful because the status rarely or never changes. So it gives membership more information about why a charge failed and it automates a process that would otherwise be manual.

#### How should this be manually tested?

- Enter a donation and use this card number: `4000000000000341`
- Then kill the server and run this in the container: `LOG_LEVEL=DEBUG python batch.py`
- Confirm that you see `Opportunity set to 'Closed Lost' with reason: Your card was declined.` in the output.

#### Have automated tests been added?

None added but some have been modified.

#### Has this been tested on mobile?

N/A

#### Are there performance implications?

No.

#### What are the relevant tickets?

https://3.basecamp.com/3098728/buckets/736178/todos/1089655685

#### How should this change be communicated to end users?

Just let Anna know.

#### Next steps?

Start charging immediately so we don't have to run the batch job as frequently.

#### Smells?

I don't think so.

#### Has the relevant documentation/wiki been updated?

Will after merge.

#### Technical debt note

None.
